p2p/transport/webrtc/udp_mux.go

```diff
@@ -0,0 +1,351 @@
+package libp2pwebrtc
+
+// udpMuxNewAddr is mostly similar to UDPMuxDefault exported from
+// pion/ice [https://github.com/pion/ice/blob/master/udp_mux.go] with
+// the only difference being the additional channel to notify the libp2p
+// transport of any STUN requests from unknown ufrags.
+
+import (
+	"errors"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pion/ice/v2"
+	"github.com/pion/logging"
+	"github.com/pion/stun"
+)
+
+type candidateAddr struct {
+	raddr *net.UDPAddr
+	ufrag string
+}
+
+var _ ice.UDPMux = &udpMuxNewAddr{}
+
+// udpMuxNewAddr is an implementation of the interface
+type udpMuxNewAddr struct {
+	params ice.UDPMuxParams
+
+	closedChan chan struct{}
+	closeOnce  sync.Once
+
+	// connsIPv4 and connsIPv6 are maps of all udpMuxedConn indexed by ufrag|network|candidateType
+	connsIPv4, connsIPv6 map[string]*udpMuxedConn
+
+	addressMapMu sync.RWMutex
+	addressMap   map[string]*udpMuxedConn
+
+	// buffer pool to recycle buffers for net.UDPAddr encodes/decodes
+	pool *sync.Pool
+
+	mu sync.Mutex
+
+	newAddrChan chan candidateAddr
+	newAddrs    map[*net.UDPAddr]struct{}
+}
+
+const maxAddrSize = 512
+const receiveMTU = 1500
+
+// NewUDPMuxNewAddr creates an implementation of UDPMux
+func NewUDPMuxNewAddr(params ice.UDPMuxParams, newAddrChan chan candidateAddr) *udpMuxNewAddr {
+	if params.Logger == nil {
+		params.Logger = logging.NewDefaultLoggerFactory().NewLogger("ice")
+	}
+
+	m := &udpMuxNewAddr{
+		addressMap: map[string]*udpMuxedConn{},
+		params:     params,
+		connsIPv4:  make(map[string]*udpMuxedConn),
+		connsIPv6:  make(map[string]*udpMuxedConn),
+		closedChan: make(chan struct{}),
+		pool: &sync.Pool{
+			New: func() interface{} {
+				// big enough buffer to fit both packet and address
+				return newBufferHolder(receiveMTU + maxAddrSize)
+			},
+		},
+		newAddrChan: newAddrChan,
+		newAddrs:    make(map[*net.UDPAddr]struct{}),
+	}
+
+	go m.connWorker()
+
+	return m
+}
+
+// LocalAddr returns the listening address of this UDPMuxNewAddr
+func (m *udpMuxNewAddr) LocalAddr() net.Addr {
+	return m.params.UDPConn.LocalAddr()
+}
+
+// GetConn returns a PacketConn given the connection's ufrag and network
+// creates the connection if an existing one can't be found
+func (m *udpMuxNewAddr) GetConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
```

Comment:
Do I understand correctly that we’re creating a new connection when we receive the STUN packet? Why do we do that?

Reply:
We either create a new connection if the ufrag is unseen or we return the existing connection. For a previously unseen STUN ufrag, the following steps are run since it could be a new client trying to connect: https://github.com/libp2p/specs/pull/412/files#diff-10725e780e3c8583b65c128ab23ebc6b3f5502f68b922e33bd86d941c7d601c3R120

Reply:
Ok, fair enough, thanks for the link.